### ISSUE TITLE
Compile errors on AIX 7.2

### ISSUE THREAD

**Issue by mgrubb** [mono/mono#9085]
## Steps to Reproduce

1. Run ./configure
2. gmake

### Current Behavior

I am getting a couple compile errors, one of which I've worked around.  The first error is related to undefined symbol for __cxa_finalize.
https://gist.github.com/mgrubb/6cd82c0a830f9c0ca982a0d7328546cd

The second issue is in runtime, building build/deps/basic-profile-check.exe:
https://gist.github.com/mgrubb/368335f564ea67e227b92bb73b5071d0

### Expected Behavior

Compiles the mono runtime successfully.

## On which platforms did you notice this

AIX 7.2

**Version Used**:

gcc (GCC) 4.8.5
libgcc 4.8.5
AIX OS Level: 7200-02-01-1731

## Stacktrace

```
Please paste the stack trace here if available.
```

**Comment by NattyNarwhal** [comment]
I don't recommend building with GCC 4.x; per David Edelsohn, there are a lot of problems with it. I build with GCC 6.4 on AIX 6.1. In particular, you will likely run into issues with finalizers on 4.x, which is what I see here.

I also work in in master (which doesn't compile yet, waiting on mono/corefx#79 to get merged) - I'll peek why.

I recommend running the runtime under GDB when dealing with issues like these.  To do so, try changing `runtime/mono-wrapper` from

```sh
exec "$r/libtool" --mode=execute "${MONO_EXECUTABLE}" --config "/home/calvin/mono-build/mono/runtime/etc/mono/config" "$@
```


to

```sh
exec "$r/libtool" --mode=execute "gdb --args ${MONO_EXECUTABLE}" --config "/home/calvin/mono-build/mono/runtime/etc/mono/config" "$@
```

You might also want to consider passing flags like `--trace` and `-v`.

**Comment by jaykrell** [comment]
This is why I don't like mono-wrapper, or wrappers in general.
I want make to output commands, that I can very easily insert gdb/lldb into.

**Comment by mgrubb** [comment]
I don't see a GCC 6.4 available on AIX, but I can find GCC 7.2.0 for AIX 7.2 or GCC 6.3.0 for AIX 7.1.  Which would you suggest I try? Can you point me to your David Edelsohn citation as my Google fu missed that.

Also thanks for the tip about the tracing.

**Comment by NattyNarwhal** [comment]
Bull provides a GCC 6.4 (and newer) RPM, but I haven't much luck with making their packages co-exist with Perzl's (haven't tried Felt's) packages, so I've had to compile GCC myself. (As for why I use 6.4, I remember Bull's packages happily using parts of libc that aren't present on PASE. You should be fine using GCC 7 or 8.)

[This](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54601) was a GCC 4 issue I had ran into that upgrading had solved.

**Comment by NattyNarwhal** [comment]
If you're also just wanting to try a build *right now*, I provide binary builds in MonoOni/binarydist that install in a self-contained `/opt/mono` prefix with all dependencies. While the name implies that it's mostly intended for i, (indeed, they're distributed in a savefile) there are also tarballs for AIX users, and those builds were built on AIX 6.1 anyways. (There hasn't been any builds in a while due to CoreFX fallout, but I'm working on getting this all merged back in and CI set up.)

**Comment by NattyNarwhal** [comment]
Ok, I looked the code:

```c
MONO_SIG_HANDLER_FUNC (, mono_sigill_signal_handler)
{
        MONO_SIG_HANDLER_INFO_TYPE *info = MONO_SIG_HANDLER_GET_INFO ();
        MONO_SIG_HANDLER_GET_CONTEXT;

        if (mono_runtime_get_no_exec ())
                exit (1);


        mono_handle_native_crash ("SIGILL", ctx, info);
        if (mono_do_crash_chaining) {
                mono_chain_signal (MONO_SIG_HANDLER_PARAMS);
                return;
        }

        g_assert_not_reached (); // +3074 for the tarball you acquired
}
```

based on the stacktrace, and SIGILL occurring some when you jump to 0x0 as a result of a missing linked in function, it's trying to jump to a non-existent function; the linker might have warned you about that. It's probably the termination handler in libgcc; again, a sign you should upgrade.

HOWEVER, it does expose another bug in signal handling (which could be new) I should investigate.

**Comment by mgrubb** [comment]
Thanks for all the information.  I did upgrade to GCC 7.2.0, and got much further in the build, this time.
Now it is complaining about a missing mscorlib.dll.

https://gist.github.com/mgrubb/6df7cf0a2f184a5e2ec4416df3a298ab

It no longer halts with the __cxa_finalize error, and I can run the runtime/_tmpinst/bin/mono command without it segfaulting.

**Comment by NattyNarwhal** [comment]
Yeah, you're at the end of the build. I'm not sure what's going on there. My suggestion: is `.` in your `$PATH`? If so, try removing it.

**Comment by mgrubb** [comment]
no `.` in my `$PATH`. It looks almost like it is either not looking in the right directory for mscorlib.dll either within the source, or it is trying to look for it in `${PREFIX}`. I do have several mscorlib.dll files in the source:
```
> find . -name mscorlib.dll
./mcs/class/lib/monolite-linux/1051200002/mscorlib.dll
./mcs/class/lib/monolite-win32/1051200002/mscorlib.dll
./mcs/class/lib/monolite-darwin/1051200002/mscorlib.dll
./external/binary-reference-assemblies/v4.0/mscorlib.dll
./external/binary-reference-assemblies/v4.6.1/mscorlib.dll
./external/binary-reference-assemblies/v2.0/mscorlib.dll
./external/binary-reference-assemblies/v4.5.2/mscorlib.dll
./external/binary-reference-assemblies/v4.6.2/mscorlib.dll
./external/binary-reference-assemblies/v4.5.1/mscorlib.dll
./external/binary-reference-assemblies/v4.7.1/mscorlib.dll
./external/binary-reference-assemblies/build/monodroid/mscorlib.dll
./external/binary-reference-assemblies/build/monotouch/mscorlib.dll
./external/binary-reference-assemblies/v4.6/mscorlib.dll
./external/binary-reference-assemblies/v4.5/mscorlib.dll
./external/binary-reference-assemblies/v4.7/mscorlib.dll
```

**Comment by NattyNarwhal** [comment]
The Monolite and refererence assemblies are acquired for bootstrap.

With `basic-profile-check` being ran, and the managed world not being built otherwise, either you've already built it, (have you seen MCS output?) or somehow, it's getting skipped and not built at all.

**Comment by NattyNarwhal** [comment]
I performed a build with AIX 6.1 fine from latest sources from master with a mix of Bull and Perzl packages. I'm trying a build on another system now, see if it's reproducible there too.

**Comment by NattyNarwhal** [comment]
Have you tried again recently, particularly from `master`? I've been building fine on another AIX 6.1 system, but with AIX Toolbox sourced GCC instead of Bull Freeware this time.

**Comment by mgrubb** [comment]
I haven't tried again from master yet.  I'll try again this week.

**Comment by edelsohn** [comment]
@NattyNarwhal Is there any Issue tracking the current AIX "JIT" work? I tried to email you directly in response to your questions to IBMers, but it bounced.

AIX is POSIX / Unix, but it is not Linux/Solaris/SVR4/ELF.  The AIX assembler syntax is markedly different.  One cannot feed it normal ELF syntax.  AIX XCOFF assembly syntax is not completely foreign and has similar concepts to ELF, but the details are different.  For example, it uses a concept called CSECTs for relocatable code sections packed into a COFF section. Some pseudo-ops have different names, like ".llong" instead of ".quad".  It doesn't use decorations for relocation types.

I would recommend that you look at the output from GCC for AIX as a guide.  Also, the AIX Assembler Language Reference

https://www.ibm.com/support/knowledgecenter/en/ssw_aix_71/com.ibm.aix.alangref/idalangref_pseudoops.htm

Also, GNU Binutils has not been maintained on AIX, so eventually you will have to switch to AIX Tools (AIX Assembelr and AIX Linker, not gas and gld).

**Comment by NattyNarwhal** [comment]
Yes, I contacted the mail admin, seems to be problems with the mailserver. I have an issue tracking what's needed for AOT [here](https://github.com/MonoOni/binarydist/issues/9) and had concluded gas was a lost cause.
